service/s3/s3crypto/authentication_mode.go

```diff
@@ -0,0 +1,50 @@
+package s3crypto
+
+import (
+	"io"
+)
+
+// AuthenticationMode will use AES GCM for the main cipher.
+type AuthenticationMode struct {
+	// Cipher will hold either a master symmetric key or a KMS client
+	// for encrypting the Envelope key
+	BaseKeyProvider
+
+	// This is the CEK algorithm used
+	CipherData
+}
+
+// Authentication returns a new encryption only mode structure with a specific cipher
+// for the master key
+func Authentication(kp KeyProvider) CryptoMode {
+	return &AuthenticationMode{
+		BaseKeyProvider: BaseKeyProvider{kp},
+		CipherData:      CipherData{"AES/GCM/NoPadding", "128"},
+	}
+}
+
+// EncryptContents will generate a random key and iv and encrypt the data using cbc
+func (mode *AuthenticationMode) EncryptContents(dst io.Writer, src io.Reader) error {
+	// Sets the key and iv to a randomly generated key and iv
+	cbc, err := NewAESGCMRandom(mode)
+	if err != nil {
+		return err
+	}
+
+	reader := cbc.Encrypt(src)
+	_, err = io.Copy(dst, reader)
+	return err
+}
+
+// DecryptContents will use the symmetric key provider to instantiate a new GCM cipher.
+// We grab a decrypt reader from gcm and wrap it in a CryptoReadCloser. The only error
+// expected here is when the key or iv is of invalid length.
+func (mode *AuthenticationMode) DecryptContents(kp KeyProvider, src io.ReadCloser) (io.ReadCloser, error) {
+	gcm, err := NewAESGCM(kp)
```

Comment:
A new mode would be created. Each mode is tied to a cipher. If a user wants to add their own cipher, they'd simply add a new mode. The reason for this design is so we didn't need some sort of enum or construction function to be passed around.